### PR TITLE
Add ConvTranspose + BatchNorm fuse pass

### PR DESCRIPTION
**Pull request by grygielski**
This PR adds conv2d_transpose + batch_norm fuse pass. 

I currently added this pass to CPUPassStrategy and MKLDNNPassStrategy but it can be also used in others.

### CODE REVIEW

**Comment by grygielski** [comment]
@luotao1 This PR should boost Face model's performance. Could you check it?

File: paddle/fluid/framework/ir/conv_bn_fuse_pass.cc [hunk 1]
```diff
@@ -92,13 +92,26 @@ void recompute_bias_and_weights(const Scope* scope,
   // Re-compute weight of conv2d from BN
   auto* weights = scope->FindVar(conv_weight->Name())->GetMutable<LoDTensor>();
   auto weights_shape = weights->dims();
-  auto weights_shape_2d = flatten_to_2d(weights_shape, 1);
+  auto weights_data = weights->mutable_data<float>(platform::CPUPlace());
+
+  // ConvTranspose weights are in IOHW format
```
Comment (jczaja) [inline paddle/fluid/framework/ir/conv_bn_fuse_pass.cc:97]: Wasn't here possible to do weights fusing using Eigen (similar like for regulr conv) ?

**Review by jczaja — DISMISSED** [review]
LGTM

File: paddle/fluid/framework/ir/conv_bn_fuse_pass.cc (hunk 1 above)
Comment (grygielski) [reply paddle/fluid/framework/ir/conv_bn_fuse_pass.cc:97]: I tried to figure out if it's possible but double data reorder will be needed in order to do that similarly to regular conv. It's because conv2d_transpose filter weights are in format IOHW what means that values that have to be multiplied by certain batch_norm calculated value aren't aligned together. To illustrate that let's assume our weight's vector is [1,1,1,1,1,1,1,1] and our batch norm values are [2,3]. For regular conv, fused weights will look like: [2,2,2,2,3,3,3,3] and for conv_transpose: [2,2,3,3,2,2,3,3] (it's just a simple example). I'm also not an expert in Eigen so there may be better solution but I can't propose such.

Comment (jczaja) [reply paddle/fluid/framework/ir/conv_bn_fuse_pass.cc:97]: ok, looks reasonable to me

**Review by jczaja — DISMISSED** [review]
LGTM

**Comment by luotao1** [comment]
> This PR should boost Face model's performance. Could you check it?

@grygielski I will check it ASAP. And please pass the pr_ci_coverage.

**Comment by grygielski** [comment]
I will create tests to satisfy CI_Coverage and post benchmarks of face model today or tomorrow.

**Comment by grygielski** [comment]
Profiling report:

Before:
```
fps: 50.1784
Event                        Calls       Total       Min.        Max.        Ave.        Ratio.      
thread0::conv2d_transpose    40004       133601      0.939064    17.029      3.3397      0.676819    
thread0::conv2d              40004       36564.1     0.321496    9.88917     0.914012    0.185232    
thread0::tanh                10001       17082.5     1.66925     3.70333     1.70808     0.0865391   
thread0::batch_norm          30003       6029.46     0.089963    1.26321     0.200962    0.030545    
thread0::relu                30003       2309.87     0.031114    0.245071    0.076988    0.0117017   
thread0::fetch               10001       1765.62     0.155819    0.470094    0.176544    0.00894456  
thread0::feed                10001       28.1434     0.001657    0.018939    0.00281405  0.000142573 
thread0::load_combine        1           14.8332     14.8332     14.8332     14.8332     7.51443e-05 
```
After:
```
fps: 52.2322
Event                        Calls       Total       Min.        Max.        Ave.        Ratio.      
thread0::conv2d_transpose    40004       133319      0.929042    21.6668     3.33263     0.702697    
thread0::conv2d              40004       36446.3     0.308973    8.99549     0.911066    0.192101    
thread0::tanh                10001       17035.7     1.67124     3.44101     1.7034      0.0897922   
thread0::fetch               10001       1772.01     0.155506    1.93444     0.177183    0.00933992  
thread0::relu                30003       1109.12     0.020214    0.254751    0.0369671   0.00584598  
thread0::feed                10001       24.5144     0.001554    0.018804    0.00245119  0.000129211 
thread0::load_combine        1           17.8336     17.8336     17.8336     17.8336     9.39975e-05
```
There is around 3% speedup since batch_norm (which execution time ratio is ~3%) is simply deleted from graph without adding additional computing complexity.

**Review by jczaja — APPROVED** [review]
I reviewed newly added UT for introduced pass . LGTM

**Comment by grygielski** [comment]
@luotao1 review this PR, please

**Comment by luotao1** [comment]
- before: 48.0023ms
```
I1012 18:41:47.405616 23697 helper.h:326] ====== batch latency: 48.0023ms, number of samples: 1, sample latency: 48.0023ms, fps: 20.8323, data type: float ======
W1012 18:41:47.406805 23697 profiler.cc:89] CUDA CUPTI is not enabled

------------------------->     Profiling Report     <-------------------------

Place: CPU
Time unit: ms
Sorted by total time in descending order in the same thread

Event                        Calls       Total       Min.        Max.        Ave.        Ratio.
thread0::conv2d_transpose    400         3795.53     2.88749     17.0744     9.48883     0.794735
thread0::conv2d              400         870.807     0.741943    3.52667     2.17702     0.182336
thread0::batch_norm          300         47.2326     0.109399    0.223445    0.157442    0.00988988
thread0::relu                300         39.0913     0.066213    0.218874    0.130304    0.00818521
thread0::tanh                100         11.7407     0.11043     0.127522    0.117407    0.00245834
thread0::fetch               100         10.9741     0.100498    0.135213    0.109741    0.00229782
thread0::feed                100         0.470948    0.003866    0.011128    0.00470948  9.86103e-05
```
- after: 47.5211ms 
```
I1012 18:53:52.037951 15407 helper.h:326] ====== batch latency: 47.5211ms, number of samples: 1, sample latency: 47.5211ms, fps: 21.0433, data type: float ======
W1012 18:53:52.039032 15407 profiler.cc:89] CUDA CUPTI is not enabled

------------------------->     Profiling Report     <-------------------------

Place: CPU
Time unit: ms
Sorted by total time in descending order in the same thread

Event                        Calls       Total       Min.        Max.        Ave.        Ratio.
thread0::conv2d_transpose    400         3808.75     2.8491      17.7729     9.52186     0.805744
thread0::conv2d              400         874.453     0.738043    3.6237      2.18613     0.184991
thread0::relu                300         21.9281     0.043042    0.146464    0.0730936   0.00463891
thread0::tanh                100         11.532      0.108986    0.148024    0.11532     0.0024396
thread0::fetch               100         9.89461     0.086688    0.121521    0.0989461   0.00209321
thread0::feed                100         0.440194    0.003368    0.013041    0.00440194  9.31235e-05
```
